### PR TITLE
Work item linking re-worked

### PR DESCRIPTION
**Pull request by sanbornsen**
- Useful methods from link service is moved to work item service.
- Link service is removed.
- Old spect for work item linking unit test is removed.
- Since, most of the template doesn't rely on component variable, need to write new logics for unit testing.
- Work item link resolver added to the work item service.
- Work item link HTML component reworked.
- Filter on avaiable link type is added.

### CODE REVIEW

File: config/karma.config.js [hunk 1]
```diff
@@ -43,6 +43,7 @@ module.exports = function (config) {
     colors: true,
     logLevel: config.LOG_INFO,
     autoWatch: false,
+    // browsers: ['Chrome', 'PhantomJS_custom'],
```
Comment (sanbornsen) [inline config/karma.config.js]: Needs to be removed

File: src/app/work-item/work-item-list/work-item-detail/work-item-link/work-item-link.component.html [hunk 2]
```diff
@@ -1,25 +1,37 @@
...
+          <span id="wi-link-total">
+            {{workItem.relationalData.totalLinkCount}}
+          </span>
+          <span class="link-header-nav-text">
+            Total Links
+          </span>
+        </div>
+        <div class="link-header-nav" *ngFor="let linkType of workItem.relationalData.linkDicts">
+            <span>
+              {{linkType.count}}
```
Comment (naina-verma) [inline src/app/work-item/work-item-list/work-item-detail/work-item-link/work-item-link.component.html]: @sanbornsen , Can you add id's for link types ,as you have added for total count. That would be great!

Comment (sanbornsen) [inline src/app/work-item/work-item-list/work-item-detail/work-item-link/work-item-link.component.html]: @naina-verma sure thing.

**Comment by nimishamukherjee** [comment]
1. Create a bug
2. Link this bug to a planner item which is not in the current view (you need to scroll a few pages to view this planner item)
3. The linked work item does not show up

File: src/app/pipes/alm-valid-link-types.pipe.ts [hunk 3]
```diff
@@ -0,0 +1,25 @@
+import { Pipe, PipeTransform } from '@angular/core';
```
Comment (sanbornsen) [inline src/app/pipes/alm-valid-link-types.pipe.ts]: We need the pipe to filter the work item link types based on current work item type. For example if current work item is a bug, it should have only 'blocks' as link type list.

**Comment by sanbornsen** [comment]
@nimishamukherjee fixed the bug.

File: src/app/pipes/alm-valid-link-types.pipe.ts [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+  transform(arr: any[], workItem: WorkItem): any {
+    let opLinkTypes: any[] = [];
+    arr.forEach((item) => {
+      if (item.relationships.source_type.data.id == workItem.relationships.baseType.data.id) {
+        opLinkTypes.push({
+          name: item.attributes['forward_name'],
+          linkId: item.id,
+          linkType: 'forward'
+        });
+      } else if (item.relationships.target_type.data.id == workItem.relationships.baseType.data.id){
```
Comment (joshuawilson) [inline src/app/pipes/alm-valid-link-types.pipe.ts]: You have a check for the Source and for the Target if they match. What happens if neither match.

File: src/app/work-item/work-item-list/work-item-detail/work-item-link/work-item-link.component.ts [hunk 5]
```diff
@@ -90,98 +85,54 @@ export class WorkItemLinkComponent implements OnInit, OnChanges {
         
     }
 
-    onSelectRelation(relation: LinkType): void{
+    onSelectRelation(relation: any): void{
```
Comment (joshuawilson) [inline src/app/work-item/work-item-list/work-item-detail/work-item-link/work-item-link.component.ts]: why is the 'any'?

Comment (sanbornsen) [inline src/app/work-item/work-item-list/work-item-detail/work-item-link/work-item-link.component.ts]: The data we are getting in this param is not an entire LinkType any more, it's some manipulated data like just the link type ID, name (forward or reverse name).

File: src/app/work-item/work-item-list/work-item-detail/work-item-link/work-item-link.component.ts [hunk 6]
```diff
@@ -90,98 +85,54 @@ export class WorkItemLinkComponent implements OnInit, OnChanges {
...
-        }).catch((error) => {
-            console.log(error);
-        });
+        this.createLinkObject(
+           this.workItem['id'], 
+           this.selectedWorkItem['id'], 
+           this.selectedLinkType.linkId, 
+           this.selectedLinkType.linkType
+          );
+        const tempValue = {'data': this.link};
```
Comment (joshuawilson) [inline src/app/work-item/work-item-list/work-item-detail/work-item-link/work-item-link.component.ts]: Why is a 'temporary value' a 'constant'? Maybe this should be a 'let'?

File: src/app/pipes/alm-valid-link-types.pipe.ts (hunk 4 above)
Comment (sanbornsen) [inline src/app/pipes/alm-valid-link-types.pipe.ts]: Well, by the logic they has to match either one of them. If they are not then it's a bug in linking.

File: src/app/work-item/work-item-list/work-item-detail/work-item-link/work-item-link.component.ts [hunk 7]
```diff
@@ -90,98 +85,54 @@ export class WorkItemLinkComponent implements OnInit, OnChanges {
...
-        .then((wItems) => {
-            let workItems = cloneDeep(wItems);
-            workItems.forEach((item: any) => {
-                this.workItemsMap[item.id] = item;
-            });
-            this.calculateTotal();
-        });
-    }
-
-    calculateTotal(): void{
```
Comment (joshuawilson) [inline src/app/work-item/work-item-list/work-item-detail/work-item-link/work-item-link.component.ts]: Why did you remove calculateTotal and updateCount?

File: src/app/work-item/work-item-list/work-item-detail/work-item-link/work-item-link.component.ts [hunk 8]
```diff
@@ -90,98 +85,54 @@ export class WorkItemLinkComponent implements OnInit, OnChanges {
...
     }
 
-    onSelectRelation(relation: LinkType): void{
+    onSelectRelation(relation: any): void{
         this.selectedLinkType = relation;
-        this.searchWorkItem.searchType = relation.relationships.target_type.data.id;
+        // this.searchWorkItem.searchType = relation.relationships.target_type.data.id;
     }
 
     createLink(event: any = null): void {
```
Comment (sanbornsen) [inline src/app/work-item/work-item-list/work-item-detail/work-item-link/work-item-link.component.ts]: Yes, my bad. Over looked it.

File: src/app/work-item/work-item-list/work-item-detail/work-item-link/work-item-link.component.ts (hunk 7 above)
Comment (sanbornsen) [inline src/app/work-item/work-item-list/work-item-detail/work-item-link/work-item-link.component.ts]: Because all these data manipulation is happening in the workItemService. 
Check models/work-item.ts file.
export class RelationalData {
  creator?: User;
  comments?: Comment[];
  assignees?: User[];
  linkDicts?: LinkDict[];
  totalLinkCount?: number;
}

File: src/app/work-item/work-item-list/work-item-detail/work-item-link/work-item-link.component.ts (hunk 6 above)
Comment (sanbornsen) [inline src/app/work-item/work-item-list/work-item-detail/work-item-link/work-item-link.component.ts]: The naming is not correct here. It's parameters instead of temporary value. I'll update it.

File: src/app/pipes/alm-valid-link-types.pipe.ts (hunk 4 above)
Comment (joshuawilson) [inline src/app/pipes/alm-valid-link-types.pipe.ts]: so if you end up with a bug in the linking will this code blow up or fail gracefully?

File: src/app/work-item/work-item-list/work-item-detail/work-item-link/work-item-link.component.ts (hunk 5 above)
Comment (joshuawilson) [inline src/app/work-item/work-item-list/work-item-detail/work-item-link/work-item-link.component.ts]: Can you use a type for that? Maybe a number or string? Just asking if it is possible.

File: src/app/pipes/alm-valid-link-types.pipe.ts (hunk 4 above)
Comment (sanbornsen) [inline src/app/pipes/alm-valid-link-types.pipe.ts]: If send a buggy link, it will not blow up, it'll just not show up the link in the list. If none of these conditions passes the link will not be pushed into the list.

File: src/app/work-item/work-item-list/work-item-detail/work-item-link/work-item-link.component.ts (hunk 5 above)
Comment (sanbornsen) [inline src/app/work-item/work-item-list/work-item-detail/work-item-link/work-item-link.component.ts]: Actually it's an object and I don't have any type defined for that kind of object. I'll add one then. Let me see.
